src/blockdata/script.rs

```diff
@@ -729,114 +729,94 @@ pub enum Instruction<'a> {
 
 /// Iterator over a script returning parsed opcodes.
 pub struct Instructions<'a> {
-    data: &'a [u8],
+    data: ::core::slice::Iter<'a, u8>,
     enforce_minimal: bool,
 }
 
+impl<'a> Instructions<'a> {
+    /// Set the iterator to end so that it won't iterate any longer
+    fn kill(&mut self) {
+        let len = self.data.len();
+        self.data.nth(len.max(1) - 1);
+    }
+
+    /// takes `len` bytes long slice from iterator and returns it advancing iterator
+    /// if the iterator is not long enough [`Error::EarlyEndOfScript`] is returned and the iterator is killed
+    /// to avoid returning an infinite stream of errors.
+    fn take_slice_or_kill(&mut self, len: usize) -> Result<&'a [u8], Error> {
+        if self.data.len() >= len {
+            let slice = &self.data.as_slice()[..len];
+            self.data.nth(len.max(1) - 1);
+            Ok(slice)
```

Comment:
If `take_slice_or_kill(0)` is called then the function will return `Ok(<empty_slice>)` but the first element of `self.data` will have been consumed, that's not what we want, right? This test fails
```rust
    #[test]
    fn empty_take_slice_or_kill() {
        let orig = Instructions {
            data: [0x00, 0x01].iter(),
            enforce_minimal: false,
        };

        let mut ins = orig.clone();
        let _ = ins.take_slice_or_kill(0).unwrap();
        assert_eq!(ins, orig);
    }
```

And in case you want it here is a passing test
```rust
    #[test]
    fn multiple_take_slice_or_kill() {
        let mut ins = Instructions {
            data: [0x00, 0x01, 0x02].iter(),
            enforce_minimal: false,
        };

        let slice = ins.take_slice_or_kill(1).unwrap();
        assert_eq!(slice, &[0x00]);

        let slice = ins.take_slice_or_kill(1).unwrap();
        assert_eq!(slice, &[0x01]);

        let want = Instructions {
            data: [0x02].iter(),
            enforce_minimal: false,
        };

        assert_eq!(ins, want)
    }
```

Reply:
Wow, nice catch!